lib/logstash/codecs/auto_flush.rb

```diff
@@ -7,20 +7,48 @@ def initialize(mc, interval)
     @stopped = Concurrent::AtomicBoolean.new # false by default
   end
 
+  # def start
+  #   # can't start if pipeline is stopping
+  #   return self if stopped?
+  #   if pending?
+  #     @task.cancel
+  #     create_task
+  #   elsif finished?
+  #     create_task
+  #   # else the task is executing
+  #   end
+  #   self
+  # end
+
   def start
     # can't start if pipeline is stopping
     return self if stopped?
-    if pending?
-      @task.reset
-    elsif finished?
-      @task = Concurrent::ScheduledTask.execute(@interval) do
-        @mc.auto_flush()
-      end
-    # else the task is executing
+
+    if pending? && @task.cancel
+      create_task
+      return
+    end
+    # maybe we have a timing edge case
+    # where pending? was true but cancel failed
+    # because the task started running
+    if finished?
+      create_task
+      return
     end
+    # else the task is executing
+    # wait for task to complete
+    # flush could feasibly block on queue access
+    @task.value
+    create_task
     self
```

Comment:
Only if we design for this:
```ruby
@ivar = AutoFlush.new(self, @auto_flush_interval).start
```
instead of
```ruby
@ivar = AutoFlush.new(self, @auto_flush_interval)
@ivar.start
# or
@ivar = AutoFlush.new(self, @auto_flush_interval).tap{|af| af.start}
```
where `start` is wholly an internal mechanism that does not intend to return any meaningful value.  In these cases, I think it is fine to return self instead of nil or, in Ruby, the last value accidentally generated somewhere in the preceding code.

For instance, in the our plugins `register` should return self. e.g. `@input = plugin_builder(klass, *args).register`

see [Fluent interface](https://en.wikipedia.org/wiki/Fluent_interface)